src/Validation/ScannableURLsRestController.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * ScannableURLsRestController class.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.2
+ */
+
+namespace AmpProject\AmpWP\Validation;
+
+use WP_Error;
+use WP_REST_Controller;
+use AmpProject\AmpWP\Infrastructure\Delayed;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * Rest endpoint for fetching the scannable URLs.
+ *
+ * @since 2.2
+ * @internal
+ */
+final class ScannableURLsRestController extends WP_REST_Controller implements Delayed, Service, Registerable {
+
+	/** @var ScannableURLProvider ScannableURLProvider instance. */
+	private $scannable_url_provider;
+
+	/**
+	 * Get the action to use for registering the service.
+	 *
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'rest_api_init';
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param ScannableURLProvider $scannable_url_provider Instance of ScannableURLProvider class.
+	 */
+	public function __construct( ScannableURLProvider $scannable_url_provider ) {
+		$this->namespace              = 'amp/v1';
+		$this->rest_base              = 'scannable-urls';
+		$this->scannable_url_provider = $scannable_url_provider;
+	}
+
+	/**
+	 * Registers all routes for the controller.
+	 */
+	public function register() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_items' ],
+					'args'                => [],
+					'permission_callback' => '__return_true',
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Retrieves total unreviewed count for validation URLs and errors.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		return rest_ensure_response( $this->scannable_url_provider->get_urls() );
```

Comment:
In addition to returning the list of URLs, this could return the nonce that is required to access validation data for a given URL. This is generated via `\AMP_Validation_Manager::get_amp_validate_nonce()` (which is also exposed in WP-CLI via `wp amp validation generate-nonce`). For any frontend URL on the site, you can take that nonce and use it as the `amp_validate` query parameter (i.e. `\AMP_Validation_Manager::VALIDATE_QUERY_VAR`) and it will return the validation data (but without persisting the data). For example:

```
https://example.com/about/?amp=1&amp_validate=123456
```

Perhaps this query parameter could be added to each of the URLs returned here, rather than passing it back as a separate field requiring the client ot compose the URL. So for example:

```suggestion
		$nonce = AMP_Validation_Manager::get_amp_validate_nonce();
		return rest_ensure_response(
			array_map(
				static function ( $entry ) use ( $nonce ) {
					$entry['validate_url'] = add_query_arg(
						[
							AMP_Validation_Manager::VALIDATE_QUERY_VAR => $nonce,
						],
						$entry['validate_url']
					);
				},
				$this->scannable_url_provider->get_urls()
			)
		);
```

Then the client just needs to use the `validate_url` property to obtain the validation data. The key thing here is that no loopback request is needed which can solve some headaches (see https://github.com/ampproject/amp-wp/issues/4979.)

The missing piece here then is _persisting_ the validation results in the CPT. That's what the loopback request currently does, is fetch the data and then pass the results into `AMP_Validated_URL_Post_Type::store_validation_errors()`. But what if the initial request to fetch the validation could also result in the data being _stored_ in addition to being returned? What if in addition to the `amp_validate` query parameter, there was an `amp_store` query parameter? While this is a tiny bit of an HTTP faux pas to _persist_ data with a `GET` request, it would largely be idempotent since each request would result in the same data being returned and stored.

This is what I have in mind:

```diff
diff --git a/includes/class-amp-theme-support.php b/includes/class-amp-theme-support.php
index b36c4089c..55044c4af 100644
--- a/includes/class-amp-theme-support.php
+++ b/includes/class-amp-theme-support.php
@@ -2008,6 +2008,38 @@ public static function prepare_response( $response, $args = [] ) {
 				$data['php_fatal_error'] = $last_error;
 			}
 			$data = array_merge( $data, AMP_Validation_Manager::get_validate_response_data( $sanitization_results ) );
+
+			if ( isset( $_GET['amp_store'] ) ) {
+				$validation_errors = wp_list_pluck( $data['results'], 'error' );
+				unset( $data['results'] );
+
+				$validated_url_post_id = AMP_Validated_URL_Post_Type::store_validation_errors(
+					$validation_errors,
+					amp_get_current_url(),
+					$data
+				);
+				if ( is_wp_error( $validated_url_post_id ) ) {
+					status_header( 500 );
+
+					$result = [];
+					foreach ( $validated_url_post_id->errors as $code => $messages ) {
+						foreach ( $messages as $message ) {
+							$result[] = [
+								'code'    => $code,
+								'message' => $message,
+							];
+						}
+					}
+					return wp_json_encode( $result );
+				} else {
+					status_header( 201 );
+					$data['validated_url_post'] = [
+						'id'        => $validated_url_post_id,
+						'edit_link' => get_edit_post_link( $validated_url_post_id, 'raw' )
+					];
+				}
+			}
+
 			return wp_json_encode( $data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES );
 		}
 
diff --git a/includes/validation/class-amp-validated-url-post-type.php b/includes/validation/class-amp-validated-url-post-type.php
index 5cf9cbd2b..0e4b5097e 100644
--- a/includes/validation/class-amp-validated-url-post-type.php
+++ b/includes/validation/class-amp-validated-url-post-type.php
@@ -765,7 +765,7 @@ protected static function normalize_url_for_storage( $url ) {
 		// Query args to be removed from validated URLs.
 		$removable_query_vars = array_merge(
 			wp_removable_query_args(),
-			[ 'preview_id', 'preview_nonce', 'preview', QueryVar::NOAMP ]
+			[ 'preview_id', 'preview_nonce', 'preview', QueryVar::NOAMP, AMP_Validation_Manager::VALIDATE_QUERY_VAR, 'amp_store' ]
 		);
 
 		// Normalize query args, removing all that are not recognized or which are removable.
```


So yeah, given any URL on the frontend, if you pass the `amp_validate` query var _and_ the `amp_store` query var, then the result would be returning the validation data _and_ storing it. You can then display this data in site scanning while also using the returned `validated_url_post.edit_link` to allow the user to see the full results for a given URL scan.

Reply:
I've added this as a drafted PR: https://github.com/ampproject/amp-wp/pull/6615

Reply:
Thank you for explaining the loopback-less approach to the validation.

> Perhaps this query parameter could be added to each of the URLs returned here, rather than passing it back as a separate field requiring the client ot compose the URL.

I like the idea of constructing complete URLs on the backend.

> But what if the initial request to fetch the validation could also result in the data being stored in addition to being returned?

What if the validation results would be cached by default? I.e. a request containing a nonce would either:
- get cached validation data (if it exists and is not stale), or
- validate, store results, and get them right away.

I guess it could lead to data storage issues, though.

Reply:
Yes, they could indeed be stored by default. 

There is no scenario currently in the plugin where validation results are obtained and then are not also stored, except for one: you can call `wp amp validation check-url` and it will return the results without storing them. But otherwise, the results are always stored (e.g. in block editor and when clicking Validate URL in the admin bar). So I don't see this being a problem for data storage.

As for returning cached validation data if not stale, I think this should necessitate a query parameter to explicitly opt-in to that, like `amp_cached_if_not_stale=1`.

Reply:
The new `amp_store` query parameter to save results as they are requested, as well as a new `amp-first` query parameter to override the template mode to be Standard are both found in this new PR: https://github.com/ampproject/amp-wp/pull/6615